extension/src/time_series/pipeline.rs

```diff
@@ -182,6 +204,302 @@ pub fn lttb_pipeline_element<'p, 'e>(
     }
 }
 
+type Interval = pg_sys::Datum;
+
+pub enum DownsampleMethod {
+    Average,
+    WeightedAverage,
+    Nearest,
+}
+
+impl From<i64> for DownsampleMethod {
+    fn from(item: i64) -> Self {
+        match item {
+            1 => DownsampleMethod::Average,
+            2 => DownsampleMethod::WeightedAverage,
+            3 => DownsampleMethod::Nearest,
+            _ => panic!("Invalid downsample method")
+        }
+    }
+}
+
+impl Into<i64> for DownsampleMethod {
+    fn into(self) -> i64 {
+        match self {
+            DownsampleMethod::Average => 1,
+            DownsampleMethod::WeightedAverage => 2,
+            DownsampleMethod::Nearest => 3,
+        }
+    }
+}
+
+impl DownsampleMethod {
+    pub fn process(&self, vals: &std::vec::Vec<TSPoint>, target: i64, interval: i64) -> TSPoint {
+        match self {
+            DownsampleMethod::Average => {
+                let mut sum = 0.0;
+                for TSPoint{val, ..} in vals.iter() {
+                    sum += val;
+                }
+                TSPoint{ts: target, val: sum / vals.len() as f64}
+            }
+            DownsampleMethod::WeightedAverage => {
+                let mut sum = 0.0;
+                let mut wsum  = 0.0;
+                for TSPoint{ts, val} in vals.iter() {
+                    let weight = (ts - target).abs() as f64 / interval as f64 / 2.0;
+                    sum += val * weight;
+                    wsum += weight;
+                }
+                TSPoint{ts: target, val: sum / wsum as f64}
+            }
+            DownsampleMethod::Nearest => {
+                let mut closest = i64::MAX;
+                let mut result = 0.0;
+                for TSPoint{ts, val} in vals.iter() {
```

Comment:
can we use a binary search for this?

Reply:
Doesn't look like any of the existing binary_search implementations will match this functionality (return the matching point if present, else return the closest value, in case of a tie return the average).  We could rewrite this to be more binary search-ish, but I'm not sure it's worth the added complexity unless we expect people to be using really populated buckets here.  Leaving this as a future optimization.